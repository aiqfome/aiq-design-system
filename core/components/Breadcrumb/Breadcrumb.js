"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Breadcrumb = void 0;
const react_1 = __importDefault(require("react"));
const prop_types_1 = __importDefault(require("prop-types"));
const styled_components_1 = __importDefault(require("styled-components"));
const Flex_1 = require("../Flex");
const Link_1 = require("../Link");
const Text_1 = require("../Text");
const Icon_1 = require("../Icon");
const BreadcrumbItem_1 = require("./BreadcrumbItem");
const BreadcrumbStyled = styled_components_1.default.ul `
  list-style: none;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
`;
exports.Breadcrumb = ({ routes }) => {
    const crumbs = routes.filter(r => r.name || r.icon);
    return (react_1.default.createElement(BreadcrumbStyled, null, crumbs.length &&
        crumbs.map((crumb, index) => (react_1.default.createElement(BreadcrumbItem_1.BreadcrumbItem, { "data-testid": 'crumb-item', overlay: crumb.overlay, key: index }, crumb.path ? (react_1.default.createElement(Link_1.Link, { "data-testid": 'crumb-item-link', ml: crumb.icon ? '5px' : '0', variant: crumb.type || 'internal', href: crumb.path },
            react_1.default.createElement(Flex_1.Flex, null,
                react_1.default.createElement(Icon_1.Icon, { mr: '5px' }, crumb.icon),
                crumb.name))) : (react_1.default.createElement(react_1.default.Fragment, null,
            crumb.icon,
            react_1.default.createElement(Text_1.Text, { ml: crumb.icon ? '5px' : '0', cursor: 'default' }, crumb.name))))))));
};
exports.Breadcrumb.defaultProps = {
    routes: []
};
exports.Breadcrumb.propTypes = {
    routes: prop_types_1.default.array.isRequired
};
