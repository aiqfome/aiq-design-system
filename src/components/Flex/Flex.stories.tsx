import React from 'react'
import { withKnobs, select, text } from '@storybook/addon-knobs'

import { Flex } from './Flex'
import { Text } from '../Text'

export default {
  component: Flex,
  title: 'Flex',
  decorators: [withKnobs]
}

export const basic: React.FC = () => (
  <Flex
    justifyContent={select(
      'justifyContent',
      { FlexStart: 'flex-start', FlexEnd: 'flex-end', Center: 'center' },
      'flex-start'
    )}
    alignItems={select(
      'alignItems',
      { FlexStart: 'flex-start', FlexEnd: 'flex-end', Center: 'center' },
      'flex-start'
    )}
    flexDirection={select(
      'flexDirection',
      { Row: 'row', Column: 'column' },
      'row'
    )}
    backgroundColor={text('backgroundColor', '#fff')}
    variant={select(
      'Variant',
      { Centralized: 'centralized', FullCentralized: 'fullCentralized' },
      'centralized'
    )}
  >
    <Text>Flex!</Text>
  </Flex>
)

export const centralized: React.FC = () => (
  <Flex variant='centralized'>
    <Text>Flex!</Text>
  </Flex>
)

export const fullCentralized: React.FC = () => (
  <Flex variant='fullCentralized'>
    <Text>Flex!</Text>
  </Flex>
)
