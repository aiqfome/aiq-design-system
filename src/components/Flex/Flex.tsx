import React from 'react'
import styled, { css, DefaultTheme } from 'styled-components'
import PropTypes from 'prop-types'
import {
  color,
  ColorProps,
  space,
  SpaceProps,
  layout,
  LayoutProps,
  fontSize,
  FontSizeProps,
  fontWeight,
  FontWeightProps,
  border,
  BorderProps,
  position,
  FlexProps,
  flex,
  PositionProps,
  flexbox,
  FlexboxProps
} from 'styled-system'

export interface Props
  extends SpaceProps,
    DefaultTheme,
    LayoutProps,
    FontSizeProps,
    FontWeightProps,
    PositionProps,
    ColorProps,
    FlexProps,
    FlexboxProps,
    BorderProps {
  variant?: 'auto' | 'centralized' | 'fullCentralized'
  color?: string
  fullHeight?: boolean
  children?: any
  className?: string
  onClick?: (e: any) => void
  onDragOver?: (e: any) => void
  onDragEnter?: (e: any) => void
  onDragLeave?: (e: any) => void
  onDragEnd?: (e: any) => void
  onDrop?: (e: any) => void
  style?: any
}

const flexVariations: { [index: string]: any } = {
  auto: css``,
  centralized: css`
    justify-content: center;
    align-items: center;
  `,
  fullCentralized: css`
    justify-content: center;
    align-items: center;
    height: 100vh;
  `
}

const FlexStyled = styled.div<Props>`
  ${space}
  ${fontSize}
  ${fontWeight}
  ${border}
  ${position}
  ${color}
  ${flexbox}
  ${flex}

  display: flex;
  
  ${({ variant }) => flexVariations[variant || 'auto']}

  ${({ fullHeight }) =>
    !!fullHeight &&
    css`
      height: 100vh;
    `}

  ${layout}
`

export const Flex = React.forwardRef<HTMLInputElement, Props>(
  ({ children, color, ...props }, ref) => {
    return (
      <FlexStyled ref={ref} color={color} {...props}>
        {children}
      </FlexStyled>
    )
  }
)

Flex.displayName = 'Flex'

Flex.propTypes = {
  ref: PropTypes.func,
  color: PropTypes.string,
  onClick: PropTypes.func,
  children: PropTypes.node
}
