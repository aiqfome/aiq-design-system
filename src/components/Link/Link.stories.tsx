import React from 'react'
import { BrowserRouter as Router } from 'react-router-dom'

import { Link } from './Link'

export default {
  component: Link,
  title: 'Link'
}

export const Basic: React.FC = () => (
  <Link color='primary' href='https://aiqfome.com/'>
    Aiqfome
  </Link>
)

export const External: React.FC = () => (
  <Link color='primary' variant='external' href='https://aiqfome.com/'>
    Aiqfome
  </Link>
)

export const Internal: React.FC = () => (
  <Router>
    <Link color='primary' variant='internal' href='/food'>
      Aiqfome
    </Link>
  </Router>
)
