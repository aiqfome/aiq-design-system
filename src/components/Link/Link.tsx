import React from 'react'
import { Link as LinkRouterDom } from 'react-router-dom'

import PropTypes from 'prop-types'
import styled, { css } from 'styled-components'
import {
  color,
  ColorProps,
  space,
  SpaceProps,
  layout,
  LayoutProps,
  fontSize,
  FontSizeProps,
  fontWeight,
  FontWeightProps
} from 'styled-system'

export interface Props
  extends ColorProps,
    SpaceProps,
    LayoutProps,
    FontSizeProps,
    FontWeightProps {
  variant?: 'internal' | 'external'
  href: string
  children?: any
  color?: string
}

export interface StyledProps
  extends ColorProps,
    SpaceProps,
    LayoutProps,
    FontSizeProps,
    FontWeightProps {
  to?: string
  href?: string
  children?: any
  color?: string
}

const styledLink = css`
  ${color}
  ${space}
  ${layout}
  ${fontSize}
  ${fontWeight}

  text-decoration: none; 

  &:hover{
    cursor: pointer;
  }
`

const InternalStyled = styled(LinkRouterDom)<StyledProps>`
  ${styledLink}
`
const ExternalStyled = styled.a<StyledProps>`
  ${styledLink}
`

export const Link: React.FC<Props> = ({
  variant = 'external',
  href,
  children,
  ...props
}) => {
  if (variant === 'external') {
    return (
      <ExternalStyled href={href} {...props}>
        {children}
      </ExternalStyled>
    )
  }
  if (variant === 'internal') {
    return (
      <InternalStyled to={href} {...props}>
        {children}
      </InternalStyled>
    )
  }
  return (
    <a href={href} {...props}>
      {children}
    </a>
  )
}

Link.propTypes = {
  variant: PropTypes.oneOf(['internal', 'external']),
  href: PropTypes.string.isRequired,
  children: PropTypes.any
}
