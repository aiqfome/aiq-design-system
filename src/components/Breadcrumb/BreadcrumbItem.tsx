import React, { useState, useCallback } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { FlexProps } from 'styled-system'
import { MdArrowDropDown } from 'react-icons/md'

import { Flex } from '../Flex'

export interface PropsItem {
  children?: any
  overlay?: any
  className?: string
}

const BreadcrumbItemStyled = styled.li`
  display: flex;
  align-items: center;
  justify-content: center;
  color: ${({ theme }) => theme.colors.darkGrey};
  font-weight: 500;

  a {
    color: ${({ theme }) => theme.colors.darkGrey};
  }

  &:last-child {
    color: ${({ theme }) => theme.colors.primary};
    a {
      color: ${({ theme }) => theme.colors.primary};
    }

    &:after {
      content: '';
    }
  }

  &:after {
    margin-right: 8px;
    margin-left: 8px;
    color: ${({ theme }) => theme.colors.darkGrey};
    content: ' / ';
  }
`

const OverlayStyled = styled(Flex)`
  position: relative;
`

const OverlayContentStyled = styled(Flex)<FlexProps>`
  position: absolute;
  top: 18px;
  width: max-content;
  background: #fff;
  border-radius: 4px;
  padding: 4px 8px;
  box-shadow: 0 3px 6px -4px rgba(0, 0, 0, 0.12),
    0 6px 16px 0 rgba(0, 0, 0, 0.08), 0 9px 28px 8px rgba(0, 0, 0, 0.05);
  transition: opacity 300ms, visibility 300ms;

  &.__breadcrumb-item-overflow-hidden {
    opacity: 0;
    visibility: hidden;
  }

  &.__breadcrumb-item-overflow-visible {
    visibility: visible;
    opacity: 1;
  }
`

export const BreadcrumbItem: React.FC<PropsItem> = ({
  overlay,
  children,
  ...props
}) => {
  const [classActiveOverflow, setClassActiveOverflow] = useState(
    '__breadcrumb-item-overflow-hidden'
  )

  const handleOnMouseOver = useCallback(() => {
    setClassActiveOverflow('__breadcrumb-item-overflow-visible')
  }, [])

  const handleOnMouseOut = useCallback(() => {
    setClassActiveOverflow('__breadcrumb-item-overflow-hidden')
  }, [])

  if (overlay) {
    return (
      <BreadcrumbItemStyled
        onMouseOver={handleOnMouseOver}
        onMouseOut={handleOnMouseOut}
        {...props}
      >
        <OverlayStyled data-testid='crumb-item-overflow'>
          {children}
          <MdArrowDropDown />

          <OverlayContentStyled
            data-testid='overlay-content'
            className={classActiveOverflow}
          >
            {overlay}
          </OverlayContentStyled>
        </OverlayStyled>
      </BreadcrumbItemStyled>
    )
  }

  return <BreadcrumbItemStyled {...props}>{children}</BreadcrumbItemStyled>
}

BreadcrumbItem.propTypes = {
  children: PropTypes.any,
  overlay: PropTypes.any
}
