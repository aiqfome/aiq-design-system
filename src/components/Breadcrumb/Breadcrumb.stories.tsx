import React from 'react'
import { MdHome, MdWork } from 'react-icons/md'

import { Breadcrumb } from './Breadcrumb'
import { Link } from '../Link'
import { Flex } from '../Flex'
import { Text } from '../Text'

export default {
  component: Breadcrumb,
  title: 'Breadcrumb'
}

export const Basic: React.FC = () => (
  <Breadcrumb
    routes={[
      {
        name: 'início',
        icon: <MdHome />
      },
      {
        path: '#',
        name: 'relatório'
      },
      {
        path: '#',
        name: 'pedidos'
      }
    ]}
  />
)

export const Overlay: React.FC = () => {
  const menu = (
    <Flex flexDirection='column'>
      <Link padding='4px' href='#'>
        Item 1
      </Link>
      <Link padding='4px' href='#'>
        Item 2
      </Link>
    </Flex>
  )

  return (
    <Breadcrumb
      routes={[
        {
          name: 'início',
          icon: <MdHome />
        },
        {
          path: '#',
          overlay: menu,
          name: 'relatório'
        },
        {
          path: '#',
          name: 'pedidos'
        }
      ]}
    />
  )
}

export const WithIcons: React.FC = () => {
  return (
    <Breadcrumb
      routes={[
        {
          icon: <MdHome />
        },
        {
          icon: <MdWork />,
          name: 'áreas'
        },
        {
          path: '#',
          name: 'pedidos'
        },
        {
          name: <Text>Teste</Text>
        }
      ]}
    />
  )
}
