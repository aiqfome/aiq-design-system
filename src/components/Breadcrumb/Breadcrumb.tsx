import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { FlexProps } from 'styled-system'
import { MdArrowDropDown } from 'react-icons/md'

import { Flex } from '../Flex'
import { Link } from '../Link'
import { Text } from '../Text'
import { Icon } from '../Icon'

export interface Props {
  routes: {
    icon?: any
    name?: any
    path?: string
    type?: 'internal' | 'external'
    overlay?: any
  }[]
}

export interface PropsItem {
  children?: any
  overlay?: any
}

const BreadcrumbStyled = styled.ul`
  list-style: none;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
`

const BreadcrumbItemStyled = styled.li`
  display: flex;
  align-items: center;
  justify-content: center;
  color: ${({ theme }) => theme.colors.darkGrey};
  font-weight: 500;

  a {
    color: ${({ theme }) => theme.colors.darkGrey};
  }

  &:last-child {
    color: ${({ theme }) => theme.colors.primary};
    a {
      color: ${({ theme }) => theme.colors.primary};
    }

    &:after {
      content: '';
    }
  }

  &:after {
    margin-right: 8px;
    margin-left: 8px;
    color: ${({ theme }) => theme.colors.darkGrey};
    content: ' / ';
  }
`

const OverlayStyled = styled(Flex)`
  position: relative;
  & > div {
    opacity: 0;
    visibility: hidden;
    transition: opacity 600ms, visibility 600ms;
  }

  &:hover > div {
    visibility: visible;
    opacity: 1;
  }
`

const OverlayContentStyled = styled(Flex)<FlexProps>`
  position: absolute;
  top: 18px;
  width: max-content;
  background: #fff;
  border-radius: 4px;
  padding: 4px 8px;
  box-shadow: 0 3px 6px -4px rgba(0, 0, 0, 0.12),
    0 6px 16px 0 rgba(0, 0, 0, 0.08), 0 9px 28px 8px rgba(0, 0, 0, 0.05);
`

const BreadcrumbItem: React.FC<PropsItem> = ({ overlay, children }) => {
  if (overlay) {
    return (
      <BreadcrumbItemStyled>
        <OverlayStyled>
          {children}
          <MdArrowDropDown />

          <OverlayContentStyled>{overlay}</OverlayContentStyled>
        </OverlayStyled>
      </BreadcrumbItemStyled>
    )
  }

  return <BreadcrumbItemStyled>{children}</BreadcrumbItemStyled>
}

BreadcrumbItem.propTypes = {
  children: PropTypes.any,
  overlay: PropTypes.any
}

export const Breadcrumb: React.FC<Props> = ({ routes }) => {
  const crumbs: any[] = routes.filter(r => r.name || r.icon)

  return (
    <BreadcrumbStyled>
      {crumbs.length &&
        crumbs.map((crumb, index) => (
          <BreadcrumbItem overlay={crumb.overlay} key={index}>
            {crumb.path ? (
              <Link
                ml={crumb.icon ? '5px' : '0'}
                variant={crumb.type || 'internal'}
                href={crumb.path}
              >
                <Flex>
                  <Icon mr='5px'>{crumb.icon}</Icon>
                  {crumb.name}
                </Flex>
              </Link>
            ) : (
              <>
                {crumb.icon}
                <Text ml={crumb.icon ? '5px' : '0'} cursor='default'>
                  {crumb.name}
                </Text>
              </>
            )}
          </BreadcrumbItem>
        ))}
    </BreadcrumbStyled>
  )
}

Breadcrumb.defaultProps = {
  routes: []
}

Breadcrumb.propTypes = {
  routes: PropTypes.array.isRequired
}
