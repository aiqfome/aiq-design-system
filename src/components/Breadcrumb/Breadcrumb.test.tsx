import React from 'react'

import { MdHome } from 'react-icons/md'

import { fireEvent } from '@testing-library/react'

import { render } from '../utils/test/render'

import { Breadcrumb } from '../Breadcrumb'
import { Link } from '../Link'
import { Flex } from '../Flex'

describe('Breadcrumb', () => {
  it('should have to render without crashing', () => {
    const routes = [
      {
        name: 'início',
        icon: <MdHome />
      },
      {
        path: '#',
        name: 'relatório'
      },
      {
        path: '#',
        name: 'pedidos'
      }
    ]

    const component = render(<Breadcrumb routes={routes} />)
    expect(component).toBeTruthy()
  })

  it('should have the same number of items from the routes property', () => {
    const routes = [
      {
        name: 'início',
        icon: <MdHome />
      },
      {
        path: '#',
        name: 'relatório'
      },
      {
        path: '#',
        name: 'pedidos'
      }
    ]

    const { getAllByTestId } = render(<Breadcrumb routes={routes} />)
    expect(getAllByTestId('crumb-item').length).toBe(routes.length)
  })

  it('should show subitems when has mouse hover on item father', () => {
    const menu = (
      <Flex flexDirection='column'>
        <Link padding='4px' href='#'>
          Item 1
        </Link>
        <Link padding='4px' href='#'>
          Item 2
        </Link>
      </Flex>
    )
    const routes = [
      {
        name: 'início',
        icon: <MdHome />
      },
      {
        path: '#',
        overlay: menu,
        name: 'relatório'
      },
      {
        path: '#',
        name: 'pedidos'
      }
    ]

    const { getByTestId } = render(<Breadcrumb routes={routes} />)

    const overlayContent = getByTestId('overlay-content')
    expect(overlayContent).toHaveClass('__breadcrumb-item-overflow-hidden')

    const crumbItemOverflow = getByTestId('crumb-item-overflow')
    fireEvent.mouseOver(crumbItemOverflow)
    expect(overlayContent).toHaveClass('__breadcrumb-item-overflow-visible')

    fireEvent.mouseOut(crumbItemOverflow)
    expect(overlayContent).toHaveClass('__breadcrumb-item-overflow-hidden')
  })

  it('should have the same number of links from the routes property ', () => {
    const routes = [
      {
        name: 'início',
        icon: <MdHome />
      },
      {
        path: '#',
        name: 'relatório'
      },
      {
        path: '#',
        name: 'pedidos'
      }
    ]

    const { getAllByTestId } = render(<Breadcrumb routes={routes} />)
    expect(getAllByTestId('crumb-item-link').length).toBe(
      routes.filter(route => route.path).length
    )
  })
})
