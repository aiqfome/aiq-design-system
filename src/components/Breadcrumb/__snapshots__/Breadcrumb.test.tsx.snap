// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must match with the previous snapshot snapshot renders 1`] = `
<ul
  className="sc-AxgMl cUkAtX"
>
  <li
    className="sc-AxheI ifaXoG"
  >
    <a
      className="sc-AxiKw fkxMDs"
      href="/"
    >
      <div
        className="sc-AxjAm hTbHNe"
      >
        <div
          className="sc-AxjAm ndLDB sc-AxhUy hrdZbX"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
        </div>
        início
      </div>
    </a>
  </li>
  <li
    className="sc-AxheI ifaXoG"
  >
    <a
      className="sc-AxiKw cOvksZ"
      href="#"
    >
      <div
        className="sc-AxjAm hTbHNe"
      >
        <div
          className="sc-AxjAm ndLDB sc-AxhUy hrdZbX"
        />
        relatório
      </div>
    </a>
  </li>
  <li
    className="sc-AxheI ifaXoG"
  >
    <a
      className="sc-AxiKw cOvksZ"
      href="#"
    >
      <div
        className="sc-AxjAm hTbHNe"
      >
        <div
          className="sc-AxjAm ndLDB sc-AxhUy hrdZbX"
        />
        pedidos
      </div>
    </a>
  </li>
</ul>
`;
